services/app-web/src/gqlHelpers/fetchHealthPlanPackageWrapper.ts

```diff
@@ -0,0 +1,191 @@
+import { ApolloError } from '@apollo/client'
+import {
+    useFetchHealthPlanPackageQuery,
+    FetchHealthPlanPackageQuery,
+    HealthPlanRevision,
+    HealthPlanRevisionEdge,
+    HealthPlanPackage,
+} from '../gen/gqlClient'
+import { HealthPlanFormDataType } from '../common-code/healthPlanFormDataType'
+import { base64ToDomain } from '../common-code/proto/healthPlanFormDataProto'
+import { recordJSException } from '../otelHelpers'
+
+type FetchLoading = {
+    status: 'LOADING'
+}
+
+type FetchError = {
+    status: 'ERROR'
+    error: Error
+}
+
+type FetchSuccess = {
+    status: 'SUCCESS'
+    data: FetchHealthPlanPackageQuery
+}
+
+type FetchHealthPlanPackageResult = FetchLoading | FetchError | FetchSuccess
+
+// These apollo results are not actually supposed to be used simultaneously, so we can make a result type that only returns the correct one
+function wrapApolloResult({
+    loading,
+    error,
+    data,
+}: {
+    loading: boolean
+    error?: ApolloError
+    data: FetchHealthPlanPackageQuery | undefined
+}): FetchHealthPlanPackageResult {
+    if (loading) {
+        return {
+            status: 'LOADING',
+        }
+    }
+
+    if (error) {
+        return {
+            status: 'ERROR',
+            error: error,
+        }
+    }
+
+    if (data) {
+        return {
+            status: 'SUCCESS',
+            data,
+        }
+    }
+
+    return {
+        status: 'ERROR',
+        error: new Error('UNEXPECTED APOLLO BEHAVIOR, NO DATA'),
+    }
+}
+
+// This is ugly, but incredibly useful.
+// We want to hide our protobuf parsing from our pages. The GQL
+// query returns, several layers down, HealthPlanRevisions which have
+// a formDataProto property on it holding the encoded proto string.
+// We want our wrapper to return the identical query to graphql, but
+// with all the HealthPlanRevisions having an additional `formData`
```

Comment:
Putting this here for posterity, since we already chatted about this in Zoom: this, and the type magic it requires, made me go "hmmm," and one reason is that it might make what we return less flexible in the future.

Reply:
I also feel uncomfortable with type gymnastics we do but I feel  this isn't too much of a transgression (side eye `toDomain`). I wonder if some kind type magic will be required for anything intended to be re useable  because we have no authoritative type for the true package (submission + revisions) with that data scattered between postgres and the proto (e.g. the graphql types for metadata + the domain model types for the form data) it becomes difficult for frontend to get any type without either mashing things together or inferring types. 

I'd be curious what you mean about less flexible in the future? I didn't see that  yet want to know what you are seeing.

Reply:
The main thing I was looking at was the recursive type replacement below, which I think is caused by putting the unwrapped data back on the package itself.  What we discussed yesterday was returning the data here as its own freestanding type.